File: internal/interfaces/ws/nip98.go
package ws

import (
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"net/url"
	"strings"

	domainevent "relay/internal/domain/event"

	"github.com/nbd-wtf/go-nostr"
)

const nip98AuthKind = 27235

// verifyNip98 validates the Authorization header of r against NIP-98, with
// the payload tag required (NIP-86 makes it mandatory; generic NIP-98 only
// recommends it). It returns the authenticated event's pubkey on success.
//
// body must be the exact bytes already read from r.Body by the caller:
// verifyNip98 does not read r.Body itself, since the NIP-86 dispatcher
// needs those same bytes to parse the RPC request.
func verifyNip98(r *http.Request, body []byte, maxAgeSeconds int) (string, error) {
	const prefix = "Nostr "
	header := r.Header.Get("Authorization")
	if !strings.HasPrefix(header, prefix) {
		return "", fmt.Errorf("missing or malformed Authorization header")
	}

	decoded, err := base64.StdEncoding.DecodeString(strings.TrimPrefix(header, prefix))
	if err != nil {
		return "", fmt.Errorf("invalid base64 in Authorization header: %w", err)
	}

	var ev nostr.Event
	if err := json.Unmarshal(decoded, &ev); err != nil {
		return "", fmt.Errorf("invalid event JSON in Authorization header: %w", err)
	}

	if ev.Kind != nip98AuthKind {
		return "", fmt.Errorf("expected kind %d, got %d", nip98AuthKind, ev.Kind)
	}
	if _, err := domainevent.NewEvent(&ev); err != nil {
		return "", fmt.Errorf("signature verification failed")
	}

	age := nostr.Now() - ev.CreatedAt
	if age < 0 {
		age = -age
	}
	if int64(age) > int64(maxAgeSeconds) {
		return "", fmt.Errorf("event created_at is more than %d seconds from now", maxAgeSeconds)
	}

	var u, method, payload string
	for _, tag := range ev.Tags {
		if len(tag) < 2 {
			continue
		}
		switch tag[0] {
		case "u":
			u = tag[1]
		case "method":
			method = tag[1]
		case "payload":
			payload = tag[1]
		}
	}

	if !strings.EqualFold(method, r.Method) {
		return "", fmt.Errorf("method tag %q does not match request method %q", method, r.Method)
	}
	if !sameRequestURL(u, r) {
		return "", fmt.Errorf("u tag %q does not match request URL", u)
	}
	if payload == "" {
		return "", fmt.Errorf("missing required payload tag")
	}
	sum := sha256.Sum256(body)
	if payload != hex.EncodeToString(sum[:]) {
		return "", fmt.Errorf("payload tag does not match request body hash")
	}

	return ev.PubKey, nil
}

// sameRequestURL compares the NIP-98 u tag against the request as this
// server observed it (path and query only), tolerant of the scheme/host
// mismatch a TLS-terminating reverse proxy can introduce. A strict
// absolute-URL comparison would need a configured canonical URL to work
// behind such a proxy, which this relay does not have.
//
// An empty path is normalized to "/" on both sides: a URL with no path
// (e.g. a bare "https://relay.example.com") is conventionally equivalent
// to one with "/", and Go's own HTTP client sends an empty-path request as
// "GET / ..." on the wire — the server never actually observes an empty
// r.URL.Path for such a request, so the client-signed u tag must match
// what the server sees, not what the client originally typed.
func sameRequestURL(u string, r *http.Request) bool {
	parsed, err := url.Parse(u)
	if err != nil {
		return false
	}
	path := parsed.Path
	if path == "" {
		path = "/"
	}
	reqPath := r.URL.Path
	if reqPath == "" {
		reqPath = "/"
	}
	return path == reqPath && parsed.RawQuery == r.URL.RawQuery
}

File: frontend/server/utils/nip98.ts
import { verifyEvent } from 'nostr-tools/pure'
import type { Event as NostrEvent } from 'nostr-tools/pure'

/**
 * NIP-98 verification for the dashboard's own login endpoint.
 *
 * This mirrors backend/internal/interfaces/ws/nip98.go, which authenticates
 * the relay's NIP-86 management API, with two deliberate differences:
 *
 *   - the auth event carries a `challenge` tag holding a server-issued
 *     nonce (see server/utils/challenge.ts);
 *   - there is no `payload` tag, because the request body is empty — the
 *     event travels in the Authorization header, so it cannot contain a
 *     hash of a body that contains itself.
 */

const NIP98_AUTH_KIND = 27235
const AUTH_HEADER_PREFIX = 'Nostr '

/** Bounds the work done for an unauthenticated caller. */
const MAX_HEADER_BYTES = 8 * 1024

export interface Nip98Verified {
  pubkey: string
  challenge: string
}

export interface Nip98Expectation {
  /** Path this server observed, e.g. /admin/api/session/login. */
  path: string
  method: string
  maxAgeSeconds: number
}

export class Nip98Error extends Error {}

export function verifyNip98Header(header: string | undefined, expected: Nip98Expectation): Nip98Verified {
  if (!header || !header.startsWith(AUTH_HEADER_PREFIX)) {
    throw new Nip98Error('missing or malformed Authorization header')
  }

  const encoded = header.slice(AUTH_HEADER_PREFIX.length)
  if (encoded.length > MAX_HEADER_BYTES) {
    throw new Nip98Error('Authorization header too large')
  }

  let event: NostrEvent
  try {
    event = JSON.parse(Buffer.from(encoded, 'base64').toString('utf8'))
  } catch {
    throw new Nip98Error('invalid event JSON in Authorization header')
  }

  if (event?.kind !== NIP98_AUTH_KIND) {
    throw new Nip98Error(`expected kind ${NIP98_AUTH_KIND}`)
  }

  // verifyEvent checks the id hash and the Schnorr signature together, so a
  // tampered tag or pubkey fails here.
  if (!verifyEvent(event)) {
    throw new Nip98Error('signature verification failed')
  }

  const age = Math.abs(Math.floor(Date.now() / 1000) - event.created_at)
  if (age > expected.maxAgeSeconds) {
    throw new Nip98Error(`created_at is more than ${expected.maxAgeSeconds} seconds from now`)
  }

  const tag = (name: string): string => event.tags.find(t => t[0] === name)?.[1] ?? ''

  if (tag('method').toUpperCase() !== expected.method.toUpperCase()) {
    throw new Nip98Error('method tag does not match request method')
  }

  // Path only, like the Go implementation's sameRequestURL: Caddy terminates
  // TLS, so the scheme and host the browser signed are not what this process
  // observes.
  if (!sameRequestPath(tag('u'), expected.path)) {
    throw new Nip98Error('u tag does not match request URL')
  }

  const challenge = tag('challenge')
  if (!challenge) {
    throw new Nip98Error('missing required challenge tag')
  }

  return { pubkey: event.pubkey, challenge }
}

function sameRequestPath(u: string, observedPath: string): boolean {
  if (!u) {
    return false
  }
  let signed: URL
  try {
    signed = new URL(u)
  } catch {
    return false
  }
  const normalize = (path: string): string => (path === '' ? '/' : path)
  return normalize(signed.pathname) === normalize(observedPath) && signed.search === ''
}
